src/main/java/net/minecraftforge/client/animation/EntityAnimation.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.client.animation;
+
+import net.minecraft.world.entity.LivingEntity;
+import org.jetbrains.annotations.NotNull;
+
+public abstract class EntityAnimation<E extends LivingEntity> implements Comparable<EntityAnimation<E>>
+{
+    private final Mode mode;
+    private final Priority priority;
+
+    public EntityAnimation(Mode mode, Priority priority)
+    {
+        this.mode = mode;
+        this.priority = priority;
+    }
+
+    public EntityAnimation(Mode mode)
+    {
+        this(mode, Priority.DEFAULT);
+    }
+
+    /**
+     * Gets the mode of the animation
+     */
+    public final Mode getMode()
+    {
+        return mode;
+    }
+
+    /**
+     * Gets the priority of the animation
+     */
+    public final Priority getPriority()
+    {
+        return priority;
+    }
+
+    /**
+     * The condition to test to determine if the animation can be executed. This is where a test like
+     * holding or using a particular item, riding an entity, etc should be used. There should always
+     * be a condition and never just true.
+     *
+     * @param entity the entity the animation is being applied to
+     * @return the result of the test
+     */
+    public abstract boolean canRun(E entity);
```

Comment:
I don't see why this method needs to exist.
The doc for this method says, "There should always be a condition and never just true.", but I don't see how that would always be the case. Many animations have no playing conditions, and this method does unnecessary branching for animations that always play.

Also, modders could want to process multiple conditioned stages to an animation(s) in one `EntityAnimation` instance as a separate instance per conditioned animation stage is verbose, and the `EntityAnimation` instances are processed iteratively. If a modder wanted to do this, this method becomes weird as the structure of an `EntityAnimation` suggests conditionals should only get done in this method.

Furthermore, this method only has the `Entity` instance in its parameters, which is pretty limiting for conditional animations as modders may want to check for values in the `ModelComponent` instance and `Context` instance in the `apply()` method.

Reply:
> The doc for this method says, "There should always be a condition and never just true.", but I don't see how that would always be the case. Many animations have no playing conditions, and this method does unnecessary branching for animations that always play.

I should reword the documentation, you should be allowed to run the animation all the time if necessary. 

> Also, modders could want to process multiple conditioned stages to an animation(s) in one EntityAnimation instance as a separate instance per conditioned animation stage is verbose, and the EntityAnimation instances are processed iteratively. If a modder wanted to do this, this method becomes weird as the structure of an EntityAnimation suggests conditionals should only get done in this method.

I definitely didn't intend for this come across to modders that you need to use multiple `EntityAnimation` instances for an animation with conditional stages. Maybe rewording the method to `canStart` would clear it up since something could start but it doesn't mean it'll finish, whereas `canRun` might indicate the whole animation will be applied. The method exists for two reasons, the main one being that once an active animation is applied, it will ignore the remaining active animations.
```java
if (animation.canRun(entity))
{
    animation.apply(entity, this.root, context);
    if (animation.getMode() == EntityAnimation.Mode.ACTIVE)
    {
        break;
    }
}
```
The second is from a readability stand point. I believe it should be clear when an animation can start. Even if I removed the `canRun` method, it'd still replicate the same bahaviour by making `apply` return a boolean. It's really just a alternaltive way to write it.

> Furthermore, this method only has the Entity instance in its parameters, which is pretty limiting for conditional animations as modders may want to check for values in the ModelComponent instance and Context instance in the apply() method.

I don't think `ModelComponent` would actually be useful. The only thing you can get from `ModelComponent` is model parts. If testing if a model part exists, as a modder you'd already know since you can look at the `LayerDefinition`. As for testing the rotation/position of a model part, I don't know if any modder needs that super fine control. If they do, it can be applied in the `apply` method.

Reply:
> I should reword the documentation, you should be allowed to run the animation all the time if necessary.

Sounds good 👍 

>  The method exists for two reasons, the main one being that once an active animation is applied, it will ignore the remaining active animations.

Ah, I totally overlooked this, and yeah, it is just an alternative way to write it.

> I don't think ModelComponent would actually be useful. The only thing you can get from ModelComponent is model parts. If testing if a model part exists, as a modder you'd already know since you can look at the LayerDefinition. As for testing the rotation/position of a model part, I don't know if any modder needs that super fine control. If they do, it can be applied in the apply method.

I agree those variables would be useless in most cases, but this system should be as versatile as possible when given easy opportunities.
If a modder needed those variables for deciding if the animation can run, there is no way to prevent the iteration over the animations from ending when their animation should not get played.

Reply:
So I've reworked the system and custom data can now be passed and retrieved from `AnimationedData`. I've included this in the `ModelAnimation#canStart` method.